libunwind/src/Registers.hpp

```diff
@@ -1851,6 +1852,43 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
   void        setVectorRegister(int num, v128 value);
   static const char *getRegisterName(int num);
   void        jumpto() { __libunwind_Registers_arm64_jumpto(this); }
+#if defined(__CHERI_PURE_CAPABILITY__) &&                                      \
+    defined(_LIBUNWIND_SANDBOX_OTYPES) && defined(_LIBUNWIND_SANDBOX_HARDENED)
+  void        unsealSP(uintcap_t sealer) {
+    assert(__builtin_cheri_sealed_get(_registers.__sp) && "Value must be sealed");
+    _registers.__sp = __builtin_cheri_unseal(_registers.__sp, sealer);
+  }
+  void        unsealFP(uintcap_t sealer) {
+    assert(__builtin_cheri_sealed_get(_registers.__fp) && "Value must be sealed");
+    _registers.__fp = __builtin_cheri_unseal(_registers.__fp, sealer);
+  }
+  void        unsealCalleeSavedRegisters(uintcap_t sealer) {
+    for (auto i = 0; i < 10; ++i) {
+      uintcap_t sealedValue = getRegister(UNW_ARM64_C19 + i);
+      // FIXME: Would be nice to enforce this invariant, but right now
+      // unw_set_reg() gets called to set what ends up in private_1, and it
+      // would require breaking libunwind's public API to seal registers through
+      // that particular path, and therefore we can't assert this.
+#if 0
+      assert((!__builtin_cheri_tag_get(sealedValue) ||
+              __builtin_cheri_sealed_get(sealedValue)) &&
+             "Value must be sealed");
+#endif
+      uintcap_t unsealedValue = sealedValue;
+      if (__builtin_cheri_sealed_get(sealedValue))
+        unsealedValue = __builtin_cheri_unseal(sealedValue, sealer);
+      // If the tag gets cleared when we attempt to unseal our value, that means
+      // that we either have a capability that was sealed to begin with, and
+      // therefore we should just return it that way, or we have a sentry which
+      // we cannot unseal.
+      if (!__builtin_cheri_tag_get(unsealedValue) &&
+          __builtin_cheri_tag_get(sealedValue))
+        unsealedValue = sealedValue;
```

Comment:
Might be clearer to write this show that the second `if` is really dependent on the first `if`
```
if (__builtin_cheri_tag_get(sealedValue) && __builtin_cheri_sealed_get(sealedValue)) {
    unsealedValue = __builtin_cheri_unseal(sealedValue, sealer);
    if (!__builtin_cheri_tag_get(unsealedValue))
        unsealedValue = sealedValue;
}
```